horovod/mxnet/mpi_ops.cc

```diff
@@ -0,0 +1,335 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#include <chrono>
+#include <memory>
+#include <thread>
+
+#include "../common/operations.h"
+#include "adapter.h"
+#include "cuda_util.h"
+#include "handle_manager.h"
+#include "mpi_ops.h"
+#include "ready_event.h"
+#include "tensor_util.h"
+
+namespace horovod {
+namespace mxnet {
+
+static HandleManager handle_manager;
+
+namespace {
+
+std::string GetOpName(std::string prefix, char* name) {
+  if (name != nullptr) {
+    return prefix + "." + std::string(name);
+  }
+  return std::string();
+}
+
+std::string GetOpNameHandle(std::string prefix, const std::string& name,
+                            int handle) {
+  if (name.length() != 0) {
+    return name;
+  }
+  return prefix + ".noname." + std::to_string(handle);
+}
+
+} // namespace
+
+int DoAllreduce(NDArray* tensor, NDArray* output, const std::string& name,
+                Callback cb) {
+  ThrowIfError(common::CheckInitialized());
+
+  auto handle = handle_manager.AllocateHandle(cb);
+  auto device = TensorUtil::GetDevice(tensor);
+  auto hvd_tensor = std::make_shared<MXTensor<NDArray>>(tensor);
+  auto hvd_context = std::make_shared<MXOpContext<NDArray>>(device, output);
+  auto hvd_output = std::make_shared<MXTensor<NDArray>>(output);
+
+  auto enqueue_result =
+      EnqueueTensorAllreduce(hvd_context, hvd_tensor, hvd_output, nullptr,
+                             GetOpNameHandle("allreduce", name, handle), device,
+                             [handle](const Status& status) {
+                               handle_manager.MarkDone(handle, status);
+                               handle_manager.ExecuteCallback(handle);
+                             });
+  ThrowIfError(enqueue_result);
+
+  return handle;
+}
+
+#if HAVE_CUDA
+int DoAllreduceCudaOnCPU(NDArray* tensor, NDArray* output, std::string& name,
+                         Callback cb) {
+  ThrowIfError(common::CheckInitialized());
+  // Make async copy of input tensor to CPU tensor and record completion event.
+  auto hvd_cpu_buffer = std::make_shared<MXTemporaryBuffer<NDArray>>(
+      CPU_DEVICE_ID, tensor->dtype());
+  TensorUtil::AsyncCopyCudaToCPU(tensor, hvd_cpu_buffer->tensor());
+  auto ready_event = std::make_shared<MXReadyEvent<NDArray>>(tensor);
+
+  auto hvd_context = std::make_shared<MXOpContext<NDArray>>(
+      CPU_DEVICE_ID, hvd_cpu_buffer->tensor());
+
+  auto handle = handle_manager.AllocateHandle(cb);
+  auto enqueue_result = EnqueueTensorAllreduce(
+      hvd_context, hvd_cpu_buffer, hvd_cpu_buffer, ready_event,
+      GetOpNameHandle("allreduce", name, handle), CPU_DEVICE_ID,
+      [handle, hvd_cpu_buffer, output](const Status& status) {
+        TensorUtil::CopyCPUToCuda(hvd_cpu_buffer->tensor(), output);
+        handle_manager.MarkDone(handle, status);
+        handle_manager.ExecuteCallback(handle);
```

Comment:
Do we need this handle manager, or can we just call callback directly, perhaps with little synchronization?  Handle manager was useful in PyTorch since it was exposed to user.